design/authenticators/resource-restrictions/resource-restrictions.md

```diff
@@ -0,0 +1,355 @@
+      errors from the user, where they will define, for example, the
+      annotation `authn-azure/subscription-ir` (note the typo). In this
+      case we'd rather fail the authentication and point them to their
+      mistake rather than ignore it, and authenticate the request in a weaker
+       granularity than what is expected by the user.
+  - Validate required restrictions exist
+    - For example, in the Azure Authenticator we require hosts to have the
+     `subscription-id` and the `resource-group` annotations. In the K8s
+      Authenticator we require the `namespace` annotation.
+  - Validate mutually exclusive restrictions
+    - For example, in the Azure Authenticator we allow hosts to have
+     only one of the `user-assigned-identity` and the `system-assigned-identity`
+     annotations. In the K8s we allow hosts to have only one of
+      the annotations `deployment`, `deployment-config` & `stateful-set`.
+
+Also here we can see some generalization.
+
+These need to be defined for each authenticator.
+To simplify that, we can define a simple class to aggregate this information.
+It will allow for each restriction to define its behavior at once, and prevent
+string duplication across the code.
+
+The class will be in the format:
+
+```ruby
+class ResourceRestrictionsConstraints
+
+  attr_reader :permitted, :required, :mutually_exclusive
+
+  def initialize
+    @permitted = []
+    @required = []
+    @mutually_exclusive = {}
+  end
+
+  def add(restriction_name:, required: false, exclusive_group_name: nil)
+    @permitted << restriction_name
+    @required << restriction_name if required
+    (@mutually_exclusive[exclusive_group_name] ||= []) << restriction_name if exclusive_group_name.present?
+    self
+  end
+end
+```
+
+It will be used by each authenticator to define how to validate the configured
+restrictions. These are the ones that were defined in the policy for the host,
+regardless of the request.
+
+Example how it will look like for `authn-azure`:
+
+```ruby
+@constraints = ResourceRestrictionsConstraints.new
+  .add(restriction_name: "subscription-id",          required: true)
+  .add(restriction_name: "resource-group",           required: true)
+  .add(restriction_name: "user-assigned-identity",   exclusive_group_name: "identity")
+  .add(restriction_name: "system-assigned-identity", exclusive_group_name: "identity")
+```
+
+The general class will be defined as follows:
+
+- input
+  - A ResourceRestrictionsConstraints
+    - permitted
+    - required
+    - non-permitted combinations
+  - The `resource_restrictions` hash built in the previous step
+- output
+  - Nothing if validation succeeded, relevant error if it failed
+
+So the class signature is:
+
+```ruby
+ValidateResourceRestrictions = CommandClass.new(
+  dependencies: {
+    logger: Rails.logger
+  },
+  inputs:   %i(resource_restrictions constraints)
+) do
+
+  def call
+    validate_restrictions_are_permitted
```

Comment:
This method ensures that the annotation names are valid?

Reply:
removed.